File: Topic1_Arrays/Day2207/exercise_1.cpp
// Use nested loop to calculate all the possible cases and find the answer
// Time complexity : O(n^2)
// Space complexity: O(1) (because no extra spaces are needed)
#include <bits/stdc++.h>
class Solution {
public:
    void twoSum(vector<int>& nums, int target) {
        for (int i = 0; i < nums.size(); i++) {
            for (int j = i + 1; j < nums.size(); j++) {
                if (nums[i] + nums[j] == target) {
                    //return {i, j};
                }
            }
        }
        //return {-1, -1};
    }
};

File: Topic1_Arrays/Day2207/Exercise1.cs
//https://leetcode.com/problems/two-sum/submissions/1340665663/

/*
+ To use Dictionary
    + Key: current value
    + Value: idx of its
+ When looping every single element, I need to get target minus current value to find
whether that value exists in dictionary or not
    + If it doesn't exist => add to dictionary
    + If it exists => add to result array to return

Space complexity: O(n)
Time complexity: O(n)
*/

namespace Day2207
{
    public class Exercise1
    {
        public int[] TwoSum(int[] nums, int target)
        {
            Dictionary<int, int> myDict = new();

            int[] result = new int[2];

            for (int i = 0; i < nums.Length; ++i)
            {
                int calc = target - nums[i];
                if (myDict.ContainsKey(calc))
                {
                    result[0] = myDict[calc];
                    result[1] = i;
                    break;

                }
                else
                    myDict[nums[i]] = i;
            }

            return result;
        }
    }
}